website/static/js/scrollspy.js

```diff
@@ -0,0 +1,37 @@
+const epsilon = 10;
+let timer;
+const onScroll = () => {
+    if (timer) {  // throttle
+        return;
+    }
+    timer = setTimeout(() => {
+        let found = false;
+        const headings = document.querySelectorAll('.toc-headings > li > a');
```

Comment:
You can cache (pre-calculate) this value after `DOMContentLoaded`. You can also probably cache positions of their headers, but I'm not sure `DOMContentLoaded` waits for fonts to be loaded and rendered.

Reply:
I think we can cache positions of headers based on the `scrollHeight`. i.e., update cached positions if `scrollHeight` has been changed. (AFAIK, modern browsers support `document.documentElement.scrollHeight`)

Reply:
I've added the query cache (though if we ever load pages via AJAX, this will break).  I don't think we need to cache the positions of the headers -- these should be effectively cached by the web browser.  This is also standard for scrollspy.